File: ts/packages/core/src/forest.ts
// Cycle-safe hierarchy resolution: items that name a parent by id become a
// tree, and an edge that would close a cycle is dropped so its item stays a
// root. The rig view (parts naming parents), the world-transform chain and
// the workspace's scene tree (instances naming hosts) all need it, because
// malformed input — a hand-edited file, a not-yet-rejected manifest — must
// not hang a renderer or make both members of a cycle silently vanish.
//
// This is the LENIENT policy, and it is one policy in one place. The strict
// counterpart is `ManifestSchema`'s refinement, which rejects the same
// shapes with §11.5 diagnostics; validation and display want different
// answers and always will. What they must not have is different answers to
// "who is this part's parent" — `computeWorldTransforms` used to carry its
// own walk that resolved a self-parent by applying its transform TWICE and
// left a phantom entry in its output map for every parent that named no
// part.

export interface ForestNode<T> {
  value: T;
  children: ForestNode<T>[];
}

// Why an item's declared parent was not used.
export type DroppedEdge = 'self' | 'unknown' | 'cycle';

export interface Hierarchy {
  // Effective parent per id: absent from the map, or undefined, when the
  // item is a root. Every chain of these terminates.
  parentOf: ReadonlyMap<string, string | undefined>;
  // Ids whose declared parent was dropped, and why. Renderers ignore this;
  // a reporting layer can say which edge it lost.
  dropped: ReadonlyMap<string, DroppedEdge>;
  // Ids in an order where every item follows its effective parent, stable
  // with respect to input order. A caller composing transforms down the
  // chain can walk this once instead of recursing.
  order: readonly string[];
}

export function resolveHierarchy<T>(
  items: readonly T[],
  idOf: (item: T) => string,
  parentOf: (item: T) => string | undefined,
): Hierarchy {
  // First id wins, matching every other by-name lookup in the codebase.
  const declared = new Map<string, string | undefined>();
  const ids: string[] = [];
  for (const item of items) {
    const id = idOf(item);
    if (declared.has(id)) continue;
    declared.set(id, parentOf(item));
    ids.push(id);
  }

  const effective = new Map<string, string | undefined>();
  const dropped = new Map<string, DroppedEdge>();
  for (const id of ids) {
    const parent = declared.get(id);
    if (parent === undefined) {
      effective.set(id, undefined);
      continue;
    }
    if (parent === id) {
      effective.set(id, undefined);
      dropped.set(id, 'self');
      continue;
    }
    if (!declared.has(parent)) {
      effective.set(id, undefined);
      dropped.set(id, 'unknown');
      continue;
    }
    // Walking up from the proposed parent must terminate; revisiting a name
    // already on the walk means this edge cannot be followed, so drop it and
    // let the item be a root.
    //
    // The walk reads DECLARED parents, never the effective ones decided so
    // far, which is what makes the answer independent of input order: every
    // member of a cycle reaches the same verdict about its own edge, so all
    // of them become roots and reversing the list changes nothing. An
    // earlier comment here claimed the opposite; the tests
    // (`rig-transform.test.ts`, "makes every member of a cycle a root,
    // symmetrically") assert the order-independence directly.
    //
    // A chain LEADING INTO a cycle is dropped too, and reported as `cycle`
    // though its own edge closes nothing — the walk from it never
    // terminates either. The whole subtree above a bad edge detaches.
    const seen = new Set<string>([id]);
    let cur: string | undefined = parent;
    let closes = false;
    while (cur !== undefined) {
      if (seen.has(cur)) {
        closes = true;
        break;
      }
      seen.add(cur);
      cur = declared.get(cur);
    }
    if (closes) {
      effective.set(id, undefined);
      dropped.set(id, 'cycle');
    } else {
      effective.set(id, parent);
    }
  }

  const order: string[] = [];
  const placed = new Set<string>();
  const place = (id: string): void => {
    if (placed.has(id)) return;
    const parent = effective.get(id);
    if (parent !== undefined) place(parent);
    placed.add(id);
    order.push(id);
  };
  for (const id of ids) place(id);

  return { parentOf: effective, dropped, order };
}

export function buildForest<T>(
  items: readonly T[],
  idOf: (item: T) => string,
  parentOf: (item: T) => string | undefined,
): ForestNode<T>[] {
  const { parentOf: effective } = resolveHierarchy(items, idOf, parentOf);
  const nodes = new Map<string, ForestNode<T>>();
  for (const item of items) nodes.set(idOf(item), { value: item, children: [] });

  const roots: ForestNode<T>[] = [];
  for (const item of items) {
    const id = idOf(item);
    const node = nodes.get(id)!;
    const parent = effective.get(id);
    if (parent === undefined) roots.push(node);
    else nodes.get(parent)!.children.push(node);
  }
  return roots;
}

File: csharp/Cuboidy/Forest.cs
// Port of ts/packages/core/src/forest.ts.
//
// `buildForest` and `ForestNode` are NOT ported. They turn the hierarchy into
// a tree of nodes for a display layer — the editor's rig view and the
// workspace's scene tree. A consumer of this library that wants a tree builds
// it from `Hierarchy.ParentOf` and `Hierarchy.Order`, which is what `Order`
// exists for: every item follows its effective parent, so one pass places
// each node under a parent that already exists.

using System;
using System.Collections.Generic;

namespace Cuboidy;

// Why an item's declared parent was not used.
public enum DroppedEdge
{
    Self,
    Unknown,
    Cycle,
}

public sealed class Hierarchy
{
    internal Hierarchy(
        IReadOnlyDictionary<string, string?> parentOf,
        IReadOnlyDictionary<string, DroppedEdge> dropped,
        IReadOnlyList<string> order)
    {
        ParentOf = parentOf;
        Dropped = dropped;
        Order = order;
    }

    // Effective parent per id: null when the item is a root. Every chain of
    // these terminates.
    //
    // TypeScript's map holds `string | undefined` and its comment allows an id
    // to be absent OR present-with-undefined; this one always has an entry per
    // id, with a null value for a root, which is the same answer for every
    // caller (both spell "no parent").
    public IReadOnlyDictionary<string, string?> ParentOf { get; }

    // Ids whose declared parent was dropped, and why. Renderers ignore this;
    // a reporting layer can say which edge it lost.
    public IReadOnlyDictionary<string, DroppedEdge> Dropped { get; }

    // Ids in an order where every item follows its effective parent, stable
    // with respect to input order. A caller composing transforms down the
    // chain can walk this once instead of recursing.
    public IReadOnlyList<string> Order { get; }
}

// Cycle-safe hierarchy resolution: items that name a parent by id become a
// tree, and an edge that would close a cycle is dropped so its item stays a
// root. The rig view (parts naming parents), the world-transform chain and
// the workspace's scene tree (instances naming hosts) all need it, because
// malformed input — a hand-edited file, a not-yet-rejected manifest — must
// not hang a renderer or make both members of a cycle silently vanish.
//
// This is the LENIENT policy, and it is one policy in one place. The strict
// counterpart is the manifest reader's §11.5 refusal; validation and display
// want different answers and always will. What they must not have is
// different answers to "who is this part's parent".
public static class Forest
{
    public static Hierarchy ResolveHierarchy<T>(
        IReadOnlyList<T> items,
        Func<T, string> idOf,
        Func<T, string?> parentOf)
    {
        if (items is null) throw new ArgumentNullException(nameof(items));
        if (idOf is null) throw new ArgumentNullException(nameof(idOf));
        if (parentOf is null) throw new ArgumentNullException(nameof(parentOf));

        // First id wins, matching every other by-name lookup in the codebase.
        var declared = new Dictionary<string, string?>(StringComparer.Ordinal);
        var ids = new List<string>();
        foreach (T item in items)
        {
            string id = idOf(item);
            if (declared.ContainsKey(id)) continue;
            declared[id] = parentOf(item);
            ids.Add(id);
        }

        var effective = new Dictionary<string, string?>(StringComparer.Ordinal);
        var dropped = new Dictionary<string, DroppedEdge>(StringComparer.Ordinal);
        foreach (string id in ids)
        {
            string? parent = declared[id];
            if (parent is null)
            {
                effective[id] = null;
                continue;
            }

            if (string.Equals(parent, id, StringComparison.Ordinal))
            {
                effective[id] = null;
                dropped[id] = DroppedEdge.Self;
                continue;
            }

            if (!declared.ContainsKey(parent))
            {
                effective[id] = null;
                dropped[id] = DroppedEdge.Unknown;
                continue;
            }

            // Walking up from the proposed parent must terminate; revisiting a
            // name already on the walk means this edge cannot be followed, so
            // drop it and let the item be a root.
            //
            // The walk reads DECLARED parents, never the effective ones decided
            // so far, which is what makes the answer independent of input
            // order: every member of a cycle reaches the same verdict about its
            // own edge, so all of them become roots and reversing the list
            // changes nothing.
            //
            // A chain LEADING INTO a cycle is dropped too, and reported as
            // `Cycle` though its own edge closes nothing — the walk from it
            // never terminates either. The whole subtree above a bad edge
            // detaches.
            var seen = new HashSet<string>(StringComparer.Ordinal) { id };
            string? cur = parent;
            bool closes = false;
            while (cur is not null)
            {
                if (seen.Contains(cur))
                {
                    closes = true;
                    break;
                }

                seen.Add(cur);
                declared.TryGetValue(cur, out cur);
            }

            if (closes)
            {
                effective[id] = null;
                dropped[id] = DroppedEdge.Cycle;
            }
            else
            {
                effective[id] = parent;
            }
        }

        var order = new List<string>();
        var placed = new HashSet<string>(StringComparer.Ordinal);

        void Place(string id)
        {
            if (placed.Contains(id)) return;
            string? parent = effective[id];
            if (parent is not null) Place(parent);
            placed.Add(id);
            order.Add(id);
        }

        foreach (string id in ids) Place(id);

        return new Hierarchy(effective, dropped, order);
    }
}
